### PR TITLE
T sl30 migration1

### PR DESCRIPTION
**Pull request by ylavoie**
This PR bring 2 new possibilities for migration:
1. Edition of multiple fields per record, either as selects or inputs
2. Allow inserting 1 simple record in an empty table. Business is a good example. SQL-Ledger has tables which are optional but mandatory in LedgerSMB. This eases the migration process.

It generalize the migration code by using the proper column id instead of the previous fixed value.
It also add back a default SQL extension to the upgrade file.

### CODE REVIEW

File: lib/LedgerSMB/Scripts/setup.pm [hunk 1]
```diff
@@ -738,33 +744,43 @@ sub _failed_check {
...
     my $hiddens = {table => $check->table,
-                    edit => $check->column,
-                           id_column => $check->{id_column},
-                            id_where => $check->{id_where},
+               id_column => $check->{id_column},
+                id_where => $check->{id_where},
+                  insert => $check->{insert},
                 database => $request->{database}};
+    my $i = 1;
+    # Move around Can't use string "ARRAY as an ARRAY ref while "strict refs" in use
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: This comment is a complete mystery to me... Are you sure it's both still relevant and the best way to document the code below it?
(Note: to me, the code below looks pretty straight forward...)

File: lib/LedgerSMB/Scripts/setup.pm [hunk 2]
```diff
@@ -804,16 +820,37 @@ sub fix_tests{
...
+    my @edits;
+    my $i = 1;
+    # Move around Can't use string "ARRAY as an ARRAY ref while "strict refs" in use
+    while (defined $request->{"edit_$i"}) {
+      push @edits, $request->{"edit_$i"};
+      $i++;
+    }
+    my $sth = $request->{insert}
+      ? $request->{dbh}->prepare(
+            "INSERT INTO $table(" . join(',',@edits) . ") VALUES(" .
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: I'm pretty sure we want to quote identifiers from `@edits`, right?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 3]
```diff
@@ -804,16 +820,37 @@ sub fix_tests{
...
+                join(',',map { '?' } @edits) . ")" )
+      : $request->{dbh}->prepare(
+            "UPDATE $table SET " .
+                join(',',map {$request->{dbh}->quote_identifier($_) . " = ?"} @edits) .
+                " where $where = ?");
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: Does `$where` come from the web? (It seems to have been retrieved from `$request` a little bit above...)

If that's the case, we *must* quote_identifier()

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 4]
```diff
@@ -106,14 +106,23 @@ Repair query table to run once per result.
...
 Repair query column to get the values from
 
 =cut
 
-has selectable_values => (is => 'ro', isa => 'Str', required => 0);
+has selectable_values => (is => 'ro', isa => 'HashRef', required => 0);
+
+=item insert
+
+Allow insert instead of update. This to set defaults on a very limited subset
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: Allow, to me, means that both will be accepted if the boolean is true, which contrasts with "instead of".

Can we be more specific? Is it either/or or is it update or insert+update?

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 5]
```diff
@@ -137,7 +146,7 @@ Repair query column to run once per result
 
 =cut
 
-has column => (is => 'ro', isa => 'Str', required => 0);
+has column => (is => 'ro', isa => 'ArrayRef[Str]', required => 0);
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: should we make this a plural? `column` suggests one as it's a singular. I think the rest of the code uses plurals for arrays?

File: lib/LedgerSMB/Scripts/setup.pm [hunk 6]
```diff
@@ -1104,7 +1141,7 @@ sub process_and_run_upgrade_script {
         template => $template,
         no_auto_output => 1,
         format_options => {extension => 'sql'},
-        output_file => 'upgrade',
+        output_file => 'upgrade.sql',
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: I thought that by setting the extension in the format options, the result would be a file *with* extension. Is this a bug elsewhere (which we may need to fix where it appears?)

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 5 above)
Comment (ylavoie) [inline lib/LedgerSMB/Upgrade_Tests.pm]: Good point. Fixed.

File: lib/LedgerSMB/Scripts/setup.pm (hunk 2 above)
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: Yes they need to.

File: lib/LedgerSMB/Scripts/setup.pm (hunk 3 above)
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: We neved did, why now?

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 4 above)
Comment (ylavoie) [inline lib/LedgerSMB/Upgrade_Tests.pm]: Insert data instead of update.

File: lib/LedgerSMB/Scripts/setup.pm (hunk 3 above)
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: > Does $where come from the web? If that's the case, we *must* quote_identifier()

That's why now.

Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: A better solution would be to use the value from the Upgrade_Tests array element to which this data belongs

File: lib/LedgerSMB/Scripts/setup.pm (hunk 1 above)
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: Our Templater code doesn't work with objects in hiddens.

File: lib/LedgerSMB/Scripts/setup.pm (hunk 3 above)
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: But we are. There is no changes in $where handling.

File: lib/LedgerSMB/Scripts/setup.pm (hunk 6 above)
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: I though so too and would have prefered that but that ain't the case. I that the job of the templater?

Comment (einhverfr) [inline lib/LedgerSMB/Scripts/setup.pm]: If we are specifying an extension it should be recognised and managed by the template engine.  Taking a quick look

Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: I thought so too and would welcome it. But it should be the object of another PR for it will bring unrelated files into this one and remove the focus.

Comment (einhverfr) [inline lib/LedgerSMB/Scripts/setup.pm]: Yeah, this is supposed to be supported by LedgerSMB::Template and LedgerSMB::Template::TXT
